### PR TITLE
New Lift Code

### PR DESCRIPTION
**Pull request by lhmcgann**
PID w/o counterbalance, commands set up for testing purposes
code still needs work and LOTS of testing (see https://trello.com/c/7OOvkOJG/167-implement-pid-on-lift-encoder checklist)

### CODE REVIEW

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/Robot.java [hunk 1]
```diff
@@ -71,6 +72,20 @@ public double getConst(String key, double def) {
 			return pref.getDouble("Const/" + key, def);
 		}
 
+		@Override
+		public String getString(String key, String def) {
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/Robot.java]: As far as I can tell, this method isn't needed/used.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/SmartDashboardInterface.java [hunk 2]
```diff
@@ -12,4 +12,6 @@
 	public void putNumber(String string, double d);
 
 	public void putBoolean(String string, boolean b);
+
+	public String getString(String key, String def);
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/SmartDashboardInterface.java]: Not needed/used.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/AutoLift.java [hunk 3]
```diff
@@ -12,103 +12,105 @@
...
-    	liftController = new PIDController(Robot.getConst("LiftkP", 1), Robot.getConst("LiftkI", 0),
-				Robot.getConst("LiftkD", 0),RobotMap.liftEnc, this);
-    	
-    	desiredPos = stage;
-    }
-
-    // Called just before this Command runs the first time
-    protected void initialize() {
+	public AutoLift(LiftHeight stage, LiftInterface lift) {
+		this.lift = lift;
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/AutoLift.java]: Can't this command just use the methods of the Lift subsystem? Note that if you need access to it's PIDController, you can use [getPIDController](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/command/PIDSubsystem.html#getPIDController--)

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/LiftToPosition.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	// Called just before this Command runs the first time
+	@Override
+	protected void initialize() {
+		// There's no way to reset the Lift's PID constants after constructing it, so if
+		// you change them you must Restart Robot Code in DS.
+		double setpoint = lift.getDesiredDistFromPos(pos);
+		lift.setSetpoint(setpoint);
+		System.out.println("Target Height: " + setpoint);
+		lift.enable();
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/LiftToPosition.java]: Should always be enabled.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/LiftToPosition.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	// Called once after isFinished returns true
+	@Override
+	protected void end() {
+		lift.disable();
+		System.out.println("Disabled");
+		// keep stopLift() after disable() so they don't conflict
+		lift.stopLift();
+		System.out.println("Stopped");
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/LiftToPosition.java]: Shouldn't disable or stop. The controller needs to keep running to keep the lift at the desired position.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/MoveLiftWithPID.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		return false;
+	}
+
+	// Called once after isFinished returns true
+	@Override
+	protected void end() {
+		lift.disable();
+		// keep stopLift() after disable() so they don't conflict
+		lift.stopLift();
+	}
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/MoveLiftWithPID.java]: Shouldn't disable or stop per above comments.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/MoveLiftWithPID.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+	// Called when another command which requires one or more of the same
+	// subsystems is scheduled to run
+	@Override
+	protected void interrupted() {
+		end();
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/MoveLiftWithPID.java]: This is the same as the parent class implementation, so you don't need this.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/UpdateLiftPosition.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	protected boolean isFinished() {
+		return false;
+	}
+
+	// Called once after isFinished returns true
+	@Override
+	protected void end() {
+		lift.stopLift();
+		lift.disable();
+	}
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/UpdateLiftPosition.java]: Same as before, don't stop/disable.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/UpdateLiftPosition.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+	// Called when another command which requires one or more of the same
+	// subsystems is scheduled to run
+	@Override
+	protected void interrupted() {
+		end();
+	}
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/UpdateLiftPosition.java]: Unnecessary, same as base.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/subsystems/Lift.java [hunk 10]
```diff
@@ -83,15 +129,65 @@ public void runMotor(double output) {
...
+
+	/**
+	 * Runs the lift motors at the value that the pid loop calculated. Used
+	 * internally by PIDSubsystem.
+	 */
+	@Override
+	protected void usePIDOutput(double output) {
+		double out = output;
+		double spd = liftEnc.getRate();
+		out += convertSpdToVoltage(spd);
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/subsystems/Lift.java]: I was wrong about this. You should be able to just add a constant configurable value. It shouldn't change with speed.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/subsystems/Lift.java [hunk 11]
```diff
@@ -83,15 +129,65 @@ public void runMotor(double output) {
...
+	}
+
+	/**
+	 * Takes a value for the current lift speed and translates it to the amount of
+	 * voltage to motors need to supply.
+	 * 
+	 * @param speed
+	 *            - the current lift speed (in/s)
+	 */
+	public double convertSpdToVoltage(double speed) {
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/subsystems/Lift.java]: Per above, this shouldn't be needed.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/LiftToPosition.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+		requires(Robot.lift);
+		this.lift = lift;
+		pos = LiftHeight.toLH(goal);
+	}
+
+	// Called just before this Command runs the first time
+	@Override
+	protected void initialize() {
+		// There's no way to reset the Lift's PID constants after constructing it, so if
+		// you change them you must Restart Robot Code in DS.
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/LiftToPosition.java]: Not true. Try something like Lift.getPIDController().setPID(p, i, d).

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/subsystems/Lift.java [hunk 13]
```diff
@@ -2,39 +2,80 @@
...
+
+	private final int NUM_STAGES;
+	private final double WIGGLE_ROOM;
+	private final double SWITCH_DIST;
+	private final double SCALE_DIST;
+	private final double BAR_DIST;
+
+	public Lift() {
+		super("Lift", Robot.getConst("LiftkP", 0.1), Robot.getConst("LiftkI", 0), Robot.getConst("LiftkD", 0),
+				Robot.getConst("LiftkF", 0.1));
```
Comment (brettle) [inline Robot2018/src/org/usfirst/frc/team199/Robot2018/subsystems/Lift.java]: We should be able to calculate reasonable defaults for the above constants to minimize or eliminate tuning.

File: Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/AutoLift.java (hunk 3 above)
Comment (lhmcgann) [reply Robot2018/src/org/usfirst/frc/team199/Robot2018/commands/AutoLift.java]: Yes. It's kind of a mute point though: my intention was to not use this command at all and instead use LiftToPosition as the better, newest implementation

**Pull request closed without merging**